### PR TITLE
[#566] [#687] HTTP REST API

### PR DESCRIPTION
**Pull request by Millz0r**
Solves #566 and #687.

- All API calls go through a single handler now.
- New GET /rest/ call as an interface to get_rows database queries.
- New POST /rest/ call as an interface to compose_transaction calls.
- JSON-RPC doesn't listen on / anymore (that access route not documented anyway).
- RPC_PASSWORD config setting is no longer mandatory.
- Adds xmltodict requirement.

The API is documented here for now:
http://docs.counterpartylib.apiary.io/#

### CODE REVIEW

**Comment by adamkrellenstein** [comment]
Minor point: the ChangeLog should note that authentication for the JSON-RPC API is disabled by default and the password parameter is no longer required.

The syntax for multiple parameters looks strange to me, but maybe it's the standard way to do it? E.g. `sends/source/mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc/destination/mtQheFaSfWELRB2MyMBaiWjdDm6ux9Ezns/AND` In that case, could `source` and `mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc` be separated by something other than a `/`? Could `source/mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc` and `destination/mtQheFaSfWELRB2MyMBaiWjdDm6ux9Ezns` be separated by a third character as well?

Can we redirect `/rpc/` to `/api/`, so the docs can just list `/rpc/` and `/rest/`?

Otherwise, on a first read-through, it looks pretty good to me!

**Comment by ouziel-slama** [comment]
Some nitpicking:

1) REST url should contains resource names and ID's, and the last part can eventually be an "action". The action parameters should be in the query string. If you consider an address as a resource, we can have something like:
`/rest/source/mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc/compose?destination=mtQheFaSfWELRB2MyMBaiWjdDm6ux9Ezns&quantity=3&asset=XCP`
To be really REST, this query, to compose a transaction, should be used with GET method (POST is used when there is something written in the database). BUT in our case I think we should ALSO accept POST for privacy (POST query string can be crypted with SSL).

2)  `handle_rest_get` and `handle_rest_post` are very similar! Do you think it's possible to DRY a little bit this 2 functions ?

**Comment by Millz0r** [comment]
- I just added more ChangeLog changes.
- Redirection to /rpc/ will be easy, I'll add that today.
- Considering URL building, the more RESTful way would be to keep most data as params, however I thought it would be more readable current way. I agree it's better to stick with the standard and use params as not to surprise users at any time.
- They are indeed very similar, I'll see if I could consolidate it into one function, espcially if we are gonna support GET on both queries.

**Comment by Millz0r** [comment]
@ouziel-slama:
1) What about table-wide selects for get_rows (with no additional arguments)? For instance `/rest/balances/get`, which would return `SELECT * FROM balances`? The resource name would then be either table name or transaction type, and all other arguments would be passed as action parameters.

**Comment by ouziel-slama** [comment]
@Millz0r :
Yes, `GET /rest/balances` should return the first 1000 (the default/max limit is 1000) rows in balances table. 
Also the `filterop` should be in the query string, not in the url: `/rest/sends/source/mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc/destination/mtQheFaSfWELRB2MyMBaiWjdDm6ux9Ezns?op=AND`
In general, only things that can be considered as resource or as resource ID should be part of the URL.

**Comment by adamkrellenstein** [comment]
@ouziel-slama, shouldn't that then be `/rest/sends?source=mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc&destination=mtQheFaSfWELRB2MyMBaiWjdDm6ux9Ezns&op=AND`

@Millz0r, I think that's more readable anyway.

**Comment by ouziel-slama** [comment]
@adamkrellenstein: :-) yeah you right, if we want to be really REST the only resource we have here is the `sends` table.. so the cleanest way is:
`sends/{tx_index or tx_hash}` to get one row, and `sends?source=..&dest=..` to get rows.
... but it's very usual to see REST APIs that passes search parameters in the URL.. I'am not against the two formats of the url.. is up to you and @Millz0r ..

**Comment by adamkrellenstein** [comment]
I strongly prefer keeping the parameters out of the URL.

**Comment by Millz0r** [comment]
Changes above, here are some example URLs that work now:

```
curl -X GET "localhost:14000/rest/send/compose?source=mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc&destination=mtQheFaSfWELRB2MyMBaiWjdDm6ux9Ezns&asset=BTC&quantity=100000"

"01000000025c5ea576702be46253a239bd1aa6b2dbed1f5ace830bbaa9939e80209c76eee4000000001976a9144838d8b3588c4c7ba7c1d06f866e9b3739c6303788acffffffff4986e0b01495bcd00dcf9e5d112c386f5bf0d5d8d926b9811e3ac9d31657f902000000006951210282b886c087eb37dc8182f14ba6cc3e9485ed618b95804d44aecc17c300b585b0210378ee11c3fb97054877a809ce083db292b16d971bcdc6aa4c8f92087133729d8b21037af2e06061b54cdfe3657bbc8496d69000b822e2db0c86ccbe376346a700b83353aeffffffff02a0860100000000001976a9148d6ae8a3b381663118b4e1eff4cfc7d0954dd6ec88ac850ae011000000001976a9144838d8b3588c4c7ba7c1d06f866e9b3739c6303788ac00000000"
```

```
curl -X GET "localhost:14000/rest/balances/get?address=mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc&quantity=10000&op=OR"`

[{"asset": "SLVBARTEN", "quantity": 10000, "address": "n3PQ3R4Hc4C7FnDoYTBSfo3wuVPxXxvKRW"}, {"asset": "BIZBAZQUX", "quantity": 50, "address": "mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc"}, {"asset": "A11001756367432399000", "quantity": 10000, "address": "muv4ujHQmbW8G7YtVmAFboN8Y431ZpvQyK"}]
```

```
curl -H "Accept: application/xml" -X GET "localhost:14000/rest/balances/get?address=mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc&quantity=10000&op=OR"

<?xml version="1.0" encoding="utf-8"?>
<balances><item><asset>SLVBARTEN</asset><quantity>10000</quantity><address>n3PQ3R4Hc4C7FnDoYTBSfo3wuVPxXxvKRW</address></item><item><asset>BIZBAZQUX</asset><quantity>50</quantity><address>mn6q3dS2EnDUx3bmyWc6D4szJNVGtaR7zc</address></item><item><asset>A11001756367432399000</asset><quantity>10000</quantity><address>muv4ujHQmbW8G7YtVmAFboN8Y431ZpvQyK</address></item></balances>
```

File: counterpartylib/server.py [hunk 1]
```diff
@@ -210,6 +210,9 @@ def initialise(database_file=None, log_file=None, api_log_file=None,
     else:
         config.RPC_HOST = 'localhost'
 
+    # The web root directory for API calls, eg. localhost:14000/api/
+    config.RPC_WEBROOT = '/api/'
```
Comment (adamkrellenstein) [inline counterpartylib/server.py]: This should be changed to `/rpc/`.

File: counterpartylib/server.py [hunk 2]
```diff
@@ -240,10 +243,9 @@ def initialise(database_file=None, log_file=None, api_log_file=None,
     #  counterpartyd API RPC password
     if rpc_password:
         config.RPC_PASSWORD = rpc_password
+        config.RPC = 'http://' + urlencode(config.RPC_USER) + ':' + urlencode(config.RPC_PASSWORD) + '@' + config.RPC_HOST + ':' + str(config.RPC_PORT) + config.RPC_WEBROOT
     else:
-        raise ConfigurationError('RPC password not set. (Use configuration file or --rpc-password=PASSWORD)')
-
-    config.RPC = 'http://' + urlencode(config.RPC_USER) + ':' + urlencode(config.RPC_PASSWORD) + '@' + config.RPC_HOST + ':' + str(config.RPC_PORT)
+        config.RPC = 'http://' + config.RPC_HOST + ':' + str(config.RPC_PORT) + config.RPC_WEBROOT
```
Comment (adamkrellenstein) [inline counterpartylib/server.py]: We need to use `urlencode` here, too. (as just above)

File: counterpartylib/lib/api.py [hunk 3]
```diff
@@ -653,6 +691,101 @@ def handle_post():
...
+                return flask.Response(error, 400, mimetype='text/plain')
+
+            # Get all arguments passed via URL.
+            url_args = path_args.split('/')
+            try:
+                query_type = url_args.pop(0).lower()
+            except IndexError:
+                error = 'No query_type provided.'
+                return flask.Response(error, 400, mimetype='text/plain')
+            # Check if message type or table name are valid.
```
Comment (ouziel-slama) [inline counterpartylib/lib/api.py]: Error should be returned in Json by default.

File: counterpartylib/lib/api.py [hunk 4]
```diff
@@ -615,21 +624,50 @@ def _set_cors_headers(response):
...
             _set_cors_headers(response)
             return response
 
-        @app.route('/', methods=["POST",])
-        @app.route('/api/', methods=["POST",])
-        @auth.login_required
-        def handle_post():
-
+        def handle_api_post(request_json):
+            """Handle /API/ POST route. Call relevant get_rows/create_transaction wrapper."""
```
Comment (ouziel-slama) [inline counterpartylib/lib/api.py]: I think should be renamed to `handle_rpc_post` or `handle_rpc`.

File: counterpartylib/lib/api.py [hunk 5]
```diff
@@ -653,6 +691,101 @@ def handle_post():
...
+                except (script.AddressError, exceptions.ComposeError, exceptions.TransactionError, exceptions.BalanceError) as error:
+                    error_msg = str(error.__class__.__name__) + ': ' + str(error)
+                    return flask.Response(error_msg, 400, mimetype='text/plain')                        
+            else:
+                # Need to de-generate extra_args to pass it through.
+                query_args = dict([item for item in extra_args])
+                try:
+                    operator = query_args.pop('op')
+                except KeyError:
+                    operator = 'AND'
```
Comment (ouziel-slama) [inline counterpartylib/lib/api.py]: Can be replaced by : `query_args.pop('op', 'AND')`

**Comment by adamkrellenstein** [comment]
ACK, with the caveat that the docs say the API isn't yet finalized.

@ouziel-slama?

**Comment by adamkrellenstein** [comment]
Also, needs a rebase.

**Comment by ouziel-slama** [comment]
ACK

**Comment by adamkrellenstein** [comment]
Merged here: https://github.com/CounterpartyXCP/counterpartyd/commit/0695ac8f645411767d71b11327f323467254866e

**Pull request closed without merging**